presto-spi/src/main/java/com/facebook/presto/spi/function/FunctionDescriptor.java

```diff
@@ -0,0 +1,170 @@
+     */
+    private final boolean isTransformingInput;
+
+    /**
+     * Indicates whether the function prunes the subfields. The function prunes the subfields if and only if it does not output any subfield besides those that this
+     * function and all the nested function accessed.
+     * Example: FILTER function does not prune the subfields while ALL_MATCH function does.
+     */
+    private final boolean isPruningSubfields;
+
+    /**
+     * Set of indices of the CallExpression arguments to that all accessed subfields from the outer CallExpressions relate. Lambda subfields from outer calls will be
+     * passed only to the argument in this set while traversing the expression tree with SubfieldExtractor visitor.
+     * If <code>callArgumentsToPassOuterLambdaSubfields</code> is <code>null</code>, SubfieldExtractor will pass outer lambda subfields to all the arguments of the
+     * CallExpression associated with the current function.
+     * If <code>callArgumentsToPassOuterLambdaSubfields</code> is <code>ImmutableSet.of()</code>, SubfieldExtractor will not pass outer lambda subfields to any
+     * of the CallExpression arguments associated with the current function.
+     */
+    private final Set<Integer> callArgumentsToPassOuterLambdaSubfields;
+
+    /**
+     * Contains the transformation function that needs to be applied to the lambda subfield from outer functions.
+     * This parameter could be non-null only if function semantic requires the changes in the lambda subfields collected from outer functions and therefore
+     * <code>isTransformingInput</code> is true.
+     * If transformation succeeded, then the returned value contains the transformed set of lambda subfields. Otherwise, the function must return
+     * <code>Optional.empty()</code> value.
+     */
+    private final Function<Set<Subfield>, Optional<Set<Subfield>>> outerLambdaSubfieldsTransform;
+
+    private final List<LambdaDescriptor> lambdaDescriptors;
+
+    public static final FunctionDescriptor DEFAULT = new FunctionDescriptor(true, true, false, new ArrayList<LambdaDescriptor>(), new HashSet<Integer>(), null);
+
+    public FunctionDescriptor(
+            boolean isAccessingSubfieldsInternally, boolean isTransformingInput,
+            boolean isPruningSubfields,
+            List<LambdaDescriptor> lambdaDescriptors,
+            Set<Integer> callArgumentsToPassOuterLambdaSubfields,
+            Function<Set<Subfield>, Optional<Set<Subfield>>> outerLambdaSubfieldsTransform)
+    {
+        this.isAccessingSubfieldsInternally = isAccessingSubfieldsInternally;
+        this.isTransformingInput = isTransformingInput;
+        this.isPruningSubfields = isPruningSubfields;
+        this.lambdaDescriptors = unmodifiableList(requireNonNull(lambdaDescriptors, "lambdaDescriptors is null"));
+        this.callArgumentsToPassOuterLambdaSubfields = callArgumentsToPassOuterLambdaSubfields != null ? unmodifiableSet(callArgumentsToPassOuterLambdaSubfields) : null;
```

Comment:
When there are none, make the set empty rather than null.

Reply:
I tried to make it more generic and have semantic for both empty set and null. Empty set mean that we do not pass subfields to any of the input arguments. This behavior is used by default. `null` means that we need to pass the subfields to ALL of the input arguments. This behavior is used in `concat` function that accepts variable length list argument (varargs).

Reply:
I can make it Optional as Rohit suggests.

Reply:
Changed type to Optional.